### PR TITLE
Add support for `azurerm_virtual_desktop_scaling_plan`

### PR DESCRIPTION
**Pull request by alexwilcox9**
Adding support for AVD Scaling Plans

I've still got tests and docs to write but would appreciate if someone could have a quick read through of this and check they're happy with my implementation (especially around naming, I've followed the ARM template somewhat verbatim)

### CODE REVIEW

**Comment by amccullough84** [comment]
Good job I looked as I was about to make a pull request myself! I've already implemented acceptance tests so could pull into your branch?

Implementation and naming is effectively identical to mine, so minor tweaks to align

**Comment by alexwilcox9** [comment]
@amccullough84 that would be great, I did wonder what the odds were of someone else looking at this at the same time!

What's the best way to do this, I'm fairly new. I had a go at setting your branch as a remote and merging it with mine but it looked like it was pulling across a lot off extra stuff like the registration info resource from the other PR
Have I got the right branch?
https://github.com/amccullough84/terraform-provider-azurerm/tree/dv-scaling

**Comment by amccullough84** [comment]
@alexwilcox9 Yeah, mine had a few other bits and had dependencies on the other pull as I'd aligned the API version in all resources to 2021-09-03 - I'll add that as a separate one. I'm just cleaning up the acceptance tests to fully align with yours - had to  make a couple of tweaks as you need the role definition for autoscaling setup for the hostpool_reference to work.

I should (in theory) be able to send a pull request to your branch - just debugging an issue with role assignments in the test and will fire it over

**Comment by amccullough84** [comment]
@alexwilcox9 sent a pull request to your branch with acceptance tests, docs and a couple of minor tweaks in the resource (to fix test issues).

Acceptance tests are all clear now:

TF_ACC=1 go test -v ./internal/services/desktopvirtualization -run=TestAccVirtualDesktopScalingPlan -timeout 30m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccVirtualDesktopScalingPlan_basic
=== PAUSE TestAccVirtualDesktopScalingPlan_basic
=== RUN   TestAccVirtualDesktopScalingPlan_complete
=== PAUSE TestAccVirtualDesktopScalingPlan_complete
=== RUN   TestAccVirtualDesktopScalingPlan_update
=== PAUSE TestAccVirtualDesktopScalingPlan_update
=== RUN   TestAccVirtualDesktopScalingPlan_requiresImport
=== PAUSE TestAccVirtualDesktopScalingPlan_requiresImport
=== CONT  TestAccVirtualDesktopScalingPlan_basic
=== CONT  TestAccVirtualDesktopScalingPlan_requiresImport
=== CONT  TestAccVirtualDesktopScalingPlan_complete
=== CONT  TestAccVirtualDesktopScalingPlan_update
--- PASS: TestAccVirtualDesktopScalingPlan_basic (262.33s)
--- PASS: TestAccVirtualDesktopScalingPlan_requiresImport (442.57s)
--- PASS: TestAccVirtualDesktopScalingPlan_update (494.22s)
--- PASS: TestAccVirtualDesktopScalingPlan_complete (505.85s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/desktopvirtualization 506.020s

**Comment by amccullough84** [comment]
For transparency, I have deliberately left the hostpool_reference block out of the acceptance test configs. To add a hostpool_reference the "Windows Virtual Desktop" application needs a custom role creating and assigning to give permissions on the hostpool. Despite the fact the role is being created and assigned to application and scoped to the resource group in the acc test configuration the resource reports it does not have permission to access the hostpool. 

I've verified the exact same configuration works perfectly outside of the acceptance test, so can only presume it's some quirk of the acceptance test itself as the resource works perfectly for normal usage.

**Comment by alexwilcox9** [comment]
@amccullough84 thanks for the PR I've merged that in now after rebasing my stuff
I think you're right, the hostpool type needn't be specified when there's only one allowed value at the moment!

Re the reference block we shall see what the maintainers here think on that, I'll sort the linting test and mark this as ready
Thanks so much for your help on this!

**Review by catriona-m — CHANGES_REQUESTED** [review]
Hi @alexwilcox9, thanks for this! Looks good so far, however I see there are some failing tests at the minute:

```
------- Stdout: -------
=== RUN   TestAccVirtualDesktopHostPool_update
=== PAUSE TestAccVirtualDesktopHostPool_update
=== CONT  TestAccVirtualDesktopHostPool_update
testcase.go:110: Step 2/3 error: After applying this test step, the plan was not empty.
stdout:
Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
~ update in-place
Terraform will perform the following actions:
# azurerm_virtual_desktop_host_pool.test will be updated in-place
~ resource "azurerm_virtual_desktop_host_pool" "test" {
id                       = "/subscriptions/*******/resourceGroups/acctestRG-vdesktop-211117232024101875/providers/Microsoft.DesktopVirtualization/hostPools/acctestHPvr43e"
name                     = "acctestHPvr43e"
tags                     = {
"Purpose" = "Acceptance-Testing"
}
# (11 unchanged attributes hidden)
+ registration_info {
+ expiration_date = (known after apply)
}
}
Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccVirtualDesktopHostPool_update (179.74s)
FAIL
```

I also left a couple of minor suggestions inline.

**Comment by amccullough84** [comment]
> Hi @alexwilcox9, thanks for this! Looks good so far, however I see there are some failing tests at the minute:
> 
> ```
> ------- Stdout: -------
> === RUN   TestAccVirtualDesktopHostPool_update
> === PAUSE TestAccVirtualDesktopHostPool_update
> === CONT  TestAccVirtualDesktopHostPool_update
> testcase.go:110: Step 2/3 error: After applying this test step, the plan was not empty.
> stdout:
> Terraform used the selected providers to generate the following execution
> plan. Resource actions are indicated with the following symbols:
> ~ update in-place
> Terraform will perform the following actions:
> # azurerm_virtual_desktop_host_pool.test will be updated in-place
> ~ resource "azurerm_virtual_desktop_host_pool" "test" {
> id                       = "/subscriptions/*******/resourceGroups/acctestRG-vdesktop-211117232024101875/providers/Microsoft.DesktopVirtualization/hostPools/acctestHPvr43e"
> name                     = "acctestHPvr43e"
> tags                     = {
> "Purpose" = "Acceptance-Testing"
> }
> # (11 unchanged attributes hidden)
> + registration_info {
> + expiration_date = (known after apply)
> }
> }
> Plan: 0 to add, 1 to change, 0 to destroy.
> --- FAIL: TestAccVirtualDesktopHostPool_update (179.74s)
> FAIL
> ```
> 
> I also left a couple of minor suggestions inline.

@catriona-m the failing test seems to be a regression outside of this PR - the acceptance test passes against the 2.85 provider but fails against the 2.86 provider.

There is a lifecycle block in the acceptance test to ignore diffs on the expiration date as it's a relative date being used so would cause a perpetual diff.

```
  lifecycle {
    ignore_changes = [
      registration_info[0].expiration_date,
    ]
  }
```
This now does not work in 2.86 unless the attribute reference is changed to:

```
  lifecycle {
    ignore_changes = [
      registration_info.expiration_date,
    ]
  }
```

We could update the acceptance test, but seems like there's a regression here?

**Comment by alexwilcox9** [comment]
HI @catriona-m I've added the requested changes now, good spot!

As @amccullough84 says I'm not sure why this is now failing, I have made the change suggested anyway and hopefully if you test again that will work but might be worth some more digging

**Comment by alexwilcox9** [comment]
So by changing 
```
  lifecycle {
    ignore_changes = [
      registration_info[0].expiration_date,
    ]
  }
```
to
```
  lifecycle {
    ignore_changes = [
      registration_info[0].expiration_date,
    ]
  }
```
I started to get the following error
```
=== CONT  TestAccVirtualDesktopHostPool_update
    testcase.go:110: Step 2/3 error: Error running pre-apply refresh: exit status 1
        
        Error: Invalid operation
        
          on terraform_plugin_test.tf line 31, in resource "azurerm_virtual_desktop_host_pool" "test":
          31:       registration_info.expiration_date,
        
        Block type "registration_info" is represented by a list of objects, so it
        must be indexed using a numeric key, like .registration_info[0].
    testing_new.go:70: Error running post-test destroy, there may be dangling resources: exit status 1
        
        Error: Invalid operation
        
          on terraform_plugin_test.tf line 31, in resource "azurerm_virtual_desktop_host_pool" "test":
          31:       registration_info.expiration_date,
        
        Block type "registration_info" is represented by a list of objects, so it
        must be indexed using a numeric key, like .registration_info[0].
```


Instead I've updated the `expiration_date` to exclude the time part of the timestamp and removed the `ignore_changes` block entirely

Is that suitable for the tests?

**Comment by alexwilcox9** [comment]
Test output:
```
$ make acctests SERVICE=desktopvirtualization TESTARGS='-run=TestAccVirtualDesktop' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/desktopvirtualization -run=TestAccVirtualDesktop -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccVirtualDesktopApplicationGroup_basic
=== PAUSE TestAccVirtualDesktopApplicationGroup_basic
=== RUN   TestAccVirtualDesktopApplicationGroup_complete
=== PAUSE TestAccVirtualDesktopApplicationGroup_complete
=== RUN   TestAccVirtualDesktopApplicationGroup_update
=== PAUSE TestAccVirtualDesktopApplicationGroup_update
=== RUN   TestAccVirtualDesktopApplicationGroup_requiresImport
=== PAUSE TestAccVirtualDesktopApplicationGroup_requiresImport
=== RUN   TestAccVirtualDesktopApplication_basic
=== PAUSE TestAccVirtualDesktopApplication_basic
=== RUN   TestAccVirtualDesktopApplication_complete
=== PAUSE TestAccVirtualDesktopApplication_complete
=== RUN   TestAccVirtualDesktopApplication_update
=== PAUSE TestAccVirtualDesktopApplication_update
=== RUN   TestAccVirtualDesktopApplication_requiresImport
=== PAUSE TestAccVirtualDesktopApplication_requiresImport
=== RUN   TestAccVirtualDesktopHostPool_basic
=== PAUSE TestAccVirtualDesktopHostPool_basic
=== RUN   TestAccVirtualDesktopHostPool_complete
=== PAUSE TestAccVirtualDesktopHostPool_complete
=== RUN   TestAccVirtualDesktopHostPool_update
=== PAUSE TestAccVirtualDesktopHostPool_update
=== RUN   TestAccVirtualDesktopHostPool_requiresImport
=== PAUSE TestAccVirtualDesktopHostPool_requiresImport
=== RUN   TestAccVirtualDesktopScalingPlan_basic
=== PAUSE TestAccVirtualDesktopScalingPlan_basic
=== RUN   TestAccVirtualDesktopScalingPlan_complete
=== PAUSE TestAccVirtualDesktopScalingPlan_complete
=== RUN   TestAccVirtualDesktopScalingPlan_update
=== PAUSE TestAccVirtualDesktopScalingPlan_update
=== RUN   TestAccVirtualDesktopScalingPlan_requiresImport
=== PAUSE TestAccVirtualDesktopScalingPlan_requiresImport
=== RUN   TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_basic
=== PAUSE TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_basic
=== RUN   TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_complete
=== PAUSE TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_complete
=== RUN   TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_update
=== PAUSE TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_update
=== RUN   TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_requiresImport
=== PAUSE TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_requiresImport
=== CONT  TestAccVirtualDesktopApplicationGroup_basic
=== CONT  TestAccVirtualDesktopHostPool_requiresImport
=== CONT  TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_basic
=== CONT  TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_update
=== CONT  TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_complete
=== CONT  TestAccVirtualDesktopApplication_update
=== CONT  TestAccVirtualDesktopScalingPlan_basic
=== CONT  TestAccVirtualDesktopScalingPlan_complete
=== CONT  TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_requiresImport
=== CONT  TestAccVirtualDesktopHostPool_update
=== CONT  TestAccVirtualDesktopScalingPlan_requiresImport
=== CONT  TestAccVirtualDesktopScalingPlan_update
--- PASS: TestAccVirtualDesktopApplicationGroup_basic (83.92s)
=== CONT  TestAccVirtualDesktopHostPool_complete
--- PASS: TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_complete (86.77s)
=== CONT  TestAccVirtualDesktopHostPool_basic
--- PASS: TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_basic (87.17s)
=== CONT  TestAccVirtualDesktopApplication_requiresImport
--- PASS: TestAccVirtualDesktopHostPool_requiresImport (105.22s)
=== CONT  TestAccVirtualDesktopApplicationGroup_requiresImport
--- PASS: TestAccVirtualDesktopHostPool_complete (71.93s)
=== CONT  TestAccVirtualDesktopApplication_complete
--- PASS: TestAccVirtualDesktopHostPool_basic (69.69s)
=== CONT  TestAccVirtualDesktopApplication_basic
--- PASS: TestAccVirtualDesktopScalingPlan_basic (169.16s)
=== CONT  TestAccVirtualDesktopApplicationGroup_update
--- PASS: TestAccVirtualDesktopHostPool_update (169.35s)
=== CONT  TestAccVirtualDesktopApplicationGroup_complete
--- PASS: TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_requiresImport (169.39s)
--- PASS: TestAccVirtualDesktopApplication_update (180.43s)
--- PASS: TestAccVirtualDesktopApplication_requiresImport (97.95s)
--- PASS: TestAccVirtualDesktopWorkspaceApplicationGroupAssociation_update (185.26s)
--- PASS: TestAccVirtualDesktopScalingPlan_requiresImport (190.70s)
--- PASS: TestAccVirtualDesktopApplicationGroup_requiresImport (95.90s)
--- PASS: TestAccVirtualDesktopApplication_basic (73.50s)
--- PASS: TestAccVirtualDesktopApplication_complete (75.01s)
--- PASS: TestAccVirtualDesktopApplicationGroup_complete (73.23s)
--- PASS: TestAccVirtualDesktopApplicationGroup_update (163.23s)
--- PASS: TestAccVirtualDesktopScalingPlan_complete (420.24s)
--- PASS: TestAccVirtualDesktopScalingPlan_update (584.86s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/desktopvirtualization 584.917s
```

**Comment by amccullough84** [comment]
> Is that suitable for the test?

I'd say yes as that should be deprecated once #14134 is merged.

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @alexwilcox9 

Thanks for this PR / making those changes - I've taken a look through and left some comments inline in addition to what @catriona-m has mentioned, but if we can fix those up then this should otherwise be good to go 👍 

Thanks!

File: internal/services/desktopvirtualization/virtual_desktop_scaling_plan_resource.go
Comment (alexwilcox9) [reply internal/services/desktopvirtualization/virtual_desktop_scaling_plan_resource.go:223]: Hi @tombuildsstuff, thanks for the review
I've done some testing on this and it is possible to create a scaling plan and not assign it to any hostpool. A little pointless but possible

File: internal/services/desktopvirtualization/virtual_desktop_scaling_plan_resource.go
Comment (alexwilcox9) [reply internal/services/desktopvirtualization/virtual_desktop_scaling_plan_resource.go:228]: Similarly I've been able to remove and add hostpools without creating new scaling plans each time

File: website/docs/r/virtual_desktop_scaling_plan.html.markdown
Comment (alexwilcox9) [reply website/docs/r/virtual_desktop_scaling_plan.html.markdown]: Yeah that's a good idea, change has been made

File: website/docs/r/virtual_desktop_scaling_plan.html.markdown
Comment (alexwilcox9) [reply website/docs/r/virtual_desktop_scaling_plan.html.markdown]: Changed as above

File: website/docs/r/virtual_desktop_scaling_plan.html.markdown
Comment (alexwilcox9) [reply website/docs/r/virtual_desktop_scaling_plan.html.markdown]: Change made as suggested

File: website/docs/r/virtual_desktop_scaling_plan.html.markdown
Comment (alexwilcox9) [reply website/docs/r/virtual_desktop_scaling_plan.html.markdown]: Good spot thanks

File: internal/services/desktopvirtualization/virtual_desktop_scaling_plan_resource.go
Comment (alexwilcox9) [reply internal/services/desktopvirtualization/virtual_desktop_scaling_plan_resource.go]: Didn't even occur to me that was something you could do! I've done that now, thanks for the tip

Comment (alexwilcox9) [reply internal/services/desktopvirtualization/virtual_desktop_scaling_plan_resource.go]: @tombuildsstuff could you check how I've implemented this ask? Now getting linting errors telling me that `schema.SchemaValidateFunc is deprecated: please use SchemaValidateDiagFunc`
I can't find a `StringMatch` function that returns a `SchemaValidateDiagFunc` rather than a `SchemaValidateFunc`

Am I missing something or do I need to put something around `StringMatch` to make is a `SchemaValidateDiagFunc`

**Review by katbyte — APPROVED** [review]
Thanks @alexwilcox9 - LGTM 🌳

**Pull request merged**
